File: python/average_heart_beat.py
import requests

def fetch_runner_records(username):

    url = f"https://jsonmock.hackerrank.com/api/marathon?name={username}"
    records = []
    page = 1
    total_pages = 1 # initialize with at least one page to enter the loop

    # Fetch all pages of records

    while page <= total_pages:
        response = requests.get(url, params={'page': page})
        data = response.json()
        if 'data' in data:
            records.extend(data['data'])
            total_pages = data['total_pages']
            page += 1
        else:
            break
    return records

def calculate_average_heartbeat(records):

    if not records:
        return None
    total_avg_heartbeat = 0.0
    count = 0
    for record in records:
        if 'avgheartbeat' in record:
            total_avg_heartbeat += float(record['avgheartbeat'])
            count += 1
    if count > 0:
        return total_avg_heartbeat / count
    else:
        return None
    
def get_avg_heart_beat(username):

    records = fetch_runner_records(username)
    average_heartbeat = calculate_average_heartbeat(records)
    return round(average_heartbeat)

runner = input() # example: Rudy Collier
print(get_avg_heart_beat(runner))

File: typescript/average_heart_beat.ts
import * as readline from "readline";

interface RunnerRecord {
  avgheartbeat?: string | number;
  [key: string]: unknown;
}

interface MarathonResponse {
  data?: RunnerRecord[];
  total_pages?: number;
}

async function fetchRunnerRecords(username: string): Promise<RunnerRecord[]> {
  const url = "https://jsonmock.hackerrank.com/api/marathon";
  const records: RunnerRecord[] = [];
  let page = 1;
  let totalPages = 1; // initialize with at least one page to enter the loop

  // Fetch all pages of records
  while (page <= totalPages) {
    const response = await fetch(`${url}?name=${encodeURIComponent(username)}&page=${page}`);
    const data: MarathonResponse = await response.json();
    if (data.data) {
      records.push(...data.data);
      totalPages = data.total_pages ?? totalPages;
      page += 1;
    } else {
      break;
    }
  }
  return records;
}

function calculateAverageHeartbeat(records: RunnerRecord[]): number | null {
  if (records.length === 0) {
    return null;
  }
  let totalAvgHeartbeat = 0;
  let count = 0;
  for (const record of records) {
    if (record.avgheartbeat !== undefined) {
      totalAvgHeartbeat += Number(record.avgheartbeat);
      count += 1;
    }
  }
  return count > 0 ? totalAvgHeartbeat / count : null;
}

async function getAvgHeartBeat(username: string): Promise<number | null> {
  const records = await fetchRunnerRecords(username);
  const averageHeartbeat = calculateAverageHeartbeat(records);
  return averageHeartbeat === null ? null : Math.round(averageHeartbeat);
}

const rl = readline.createInterface({ input: process.stdin });
rl.question("", async (runner: string) => {
  // example: Rudy Collier
  console.log(await getAvgHeartBeat(runner));
  rl.close();
});
